### PR TITLE
Update CMake rules to include Nvidia Video Codec SDK encoding libs

### PR DESCRIPTION
**Pull request by cudawarped**
Update CMake rules to include the latest Nvidia Video Codec SDK the WITH_NVCUVENC=ON flag.

Dependant on https://github.com/opencv/opencv_contrib/pull/3360.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: cmake/OpenCVDetectCUDA.cmake [hunk 1]
```diff
@@ -71,19 +72,22 @@ if(CUDA_FOUND)
...
+		ocv_cuda_SEARCH_NVCUVID_HEADER("dynlink_nvcuvid.h" HAVE_DYNLINK_NVCUVID_HEADER)
+		find_cuda_helper_libs(nvcuvid)
+		if(CUDA_nvcuvid_LIBRARY AND (${HAVE_NVCUVID_HEADER} OR ${HAVE_DYNLINK_NVCUVID_HEADER}))
+		  # make sure to have both header and library before enabling
+		  set(HAVE_NVCUVID 1)
+		endif()
+	endif()
+	if(WITH_NVCUVENC)
+		ocv_cuda_SEARCH_NVCUVID_HEADER("nvEncodeAPI.h" HAVE_NVCUVENC_HEADER)
+		find_cuda_helper_libs(nvencodeapi)
```
Comment (asmorkalov) [inline cmake/OpenCVDetectCUDA.cmake:86]: It does not work for Linux. Linux version is named differently: https://docs.nvidia.com/video-technologies/video-codec-sdk/release-notes/

File: cmake/templates/OpenCVConfig-CUDA.cmake.in [hunk 2]
```diff
@@ -40,8 +40,8 @@ if(OpenCV_USE_NVCUVID)
   list(APPEND OpenCV_CUDA_LIBS_ABSPATH ${CUDA_nvcuvid_LIBRARIES})
 endif()
 
-if(WIN32)
-  list(APPEND OpenCV_CUDA_LIBS_ABSPATH ${CUDA_nvcuvenc_LIBRARIES})
+if(OpenCV_USE_NVCUVENC)
+  list(APPEND OpenCV_CUDA_LIBS_ABSPATH ${CUDA_nvencodeapi_LIBRARIES})
 endif()
```
[the remark this answers is not in the document]
Comment (cudawarped) [reply cmake/templates/OpenCVConfig-CUDA.cmake.in]: Correct removed and now only referenced inside cudacodec.

**Comment by cudawarped** [comment]
Sure thing, I need to fix the samples in this repo aswell.

File: samples/gpu/video_writer.cpp [hunk 3]
```diff
@@ -31,7 +31,8 @@ int main(int argc, const char* argv[])
     }
 
     cv::cuda::printShortCudaDeviceInfo(cv::cuda::getDevice());
-
+    String outputFileCpu = cv::tempfile(".avi");
+    String outputFileGpu = cv::tempfile(".h264");
```
Comment (asmorkalov) [inline samples/gpu/video_writer.cpp:35]: It's sample. User expects to get result in current directory and usually do not know where tmp location. I propose to revert all changes related to `cv::tempfile`

**Pull request merged**